ampd/src/sui/verifier.rs

```diff
@@ -0,0 +1,194 @@
+
+        let msg = Message {
+            tx_id: TransactionDigest::random(),
+            event_index: 999,
+            source_address: SuiAddress::random_for_testing_only(),
+            destination_chain: "ethereum".parse().unwrap(),
+            destination_address: format!("0x{:x}", EVMAddress::random()).parse().unwrap(),
+            payload_hash: Hash::random(),
+        };
+
+        let json_str = format!(
+            r#"{{"destination_address": "{}", "destination_chain": "{}",  "payload": "[1,2,3]",
+            "payload_hash": "{}",  "source_id": "{}"}}"#,
+            msg.destination_address,
+            msg.destination_chain,
+            msg.payload_hash.encode_hex(),
+            msg.source_address
+        );
+        let parsed: serde_json::Value = serde_json::from_str(json_str.as_str()).unwrap();
+
+        let event = SuiEvent {
+            id: EventID {
+                tx_digest: msg.tx_id,
+                event_seq: msg.event_index,
+            },
+            package_id: gateway_address.into(),
+            transaction_module: "gateway".parse().unwrap(),
+            sender: msg.source_address,
+            type_: StructTag {
+                address: gateway_address.into(),
+                module: "gateway".parse().unwrap(),
+                name: "ContractCall".parse().unwrap(),
+                type_params: vec![],
+            },
+            parsed_json: parsed,
+            bcs: vec![],
+            timestamp_ms: None,
+        };
+
+        let tx_block = SuiTransactionBlockResponse {
+            digest: msg.tx_id,
+            events: Some(SuiTransactionBlockEvents { data: vec![event] }),
+            ..Default::default()
+        };
+
+        (gateway_address, tx_block, msg)
+    }
+
+    #[test]
+    fn should_not_verify_msg_if_tx_id_does_not_match() {
+        let (gateway_address, tx_receipt, mut msg) = get_matching_msg_and_tx_block();
+
+        msg.tx_id = TransactionDigest::random();
+        assert!(!verify_message(&gateway_address, &tx_receipt, &msg));
+    }
+
+    #[test]
+    fn should_not_verify_msg_if_event_index_does_not_match() {
+        let (gateway_address, tx_receipt, mut msg) = get_matching_msg_and_tx_block();
+
+        msg.event_index = 0;
+        assert!(!verify_message(&gateway_address, &tx_receipt, &msg));
+    }
+
+    #[test]
+    fn should_not_verify_msg_if_source_address_does_not_match() {
+        let (gateway_address, tx_receipt, mut msg) = get_matching_msg_and_tx_block();
+
+        msg.source_address = SuiAddress::random_for_testing_only();
+        assert!(!verify_message(&gateway_address, &tx_receipt, &msg));
+    }
+
+    #[test]
+    fn should_not_verify_msg_if_destination_chain_does_not_match() {
+        let (gateway_address, tx_receipt, mut msg) = get_matching_msg_and_tx_block();
+
+        msg.destination_chain = "some-other-chain".parse().unwrap();
+        assert!(!verify_message(&gateway_address, &tx_receipt, &msg));
+    }
```

Comment:
can you make it more obvious in the setup that the values don't match, e.g. set both the expected and "actual" index inside the test